tests/ScopeTest.php

```diff
@@ -0,0 +1,319 @@
+
+        // Specifying hasMany here will perform input
+        $this->country->import([
+            ['name' => 'Canada', 'code' => 'CA'],
+            ['name' => 'Latvia', 'code' => 'LV'],
+            ['name' => 'Japan', 'code' => 'JP'],
+            ['name' => 'Lithuania', 'code' => 'LT', 'is_eu' => true],
+            ['name' => 'Russia', 'code' => 'RU'],
+            ['name' => 'France', 'code' => 'FR'],
+            ['name' => 'Brazil', 'code' => 'BR'],
+        ]);
+
+        $this->user = new SUser($this->db);
+
+        $this->getMigrator($this->user)->drop()->create();
+
+        $this->user->import([
+            ['name' => 'John', 'surname' => 'Smith', 'country_code' => 'CA'],
+            ['name' => 'Jane', 'surname' => 'Doe', 'country_code' => 'LV'],
+            ['name' => 'Alain', 'surname' => 'Prost', 'country_code' => 'FR'],
+            ['name' => 'Aerton', 'surname' => 'Senna', 'country_code' => 'BR'],
+            ['name' => 'Rubens', 'surname' => 'Barichello', 'country_code' => 'BR'],
+        ]);
+    }
+
+    public function testCondition()
+    {
+        $user = clone $this->user;
+
+        $condition = new BasicCondition('name', 'John');
+
+        $user->scope()->add($condition);
+
+        $user->loadAny();
+
+        $this->assertEquals('Smith', $user->get('surname'));
+    }
+
+    public function testContitionToWords()
+    {
+        $user = clone $this->user;
+
+        $condition = new BasicCondition(new Expression('false'));
+
+        $this->assertEquals('expression \'false\'', $condition->on($user)->toWords());
+
+        $condition = new BasicCondition('country_id/code', 'US');
+
+        $this->assertEquals('User that has reference Country Id where Code is equal to \'US\'', $condition->on($user)->toWords());
+
+        $condition = new BasicCondition('country_id', 2);
+
+        $this->assertEquals('Country Id is equal to \'Latvia\'', $condition->on($user)->toWords());
+
+        if ($this->driverType == 'sqlite') {
+            $condition = new BasicCondition('name', $user->expr('[surname]'));
+
+            $this->assertEquals('Name is equal to expression \'"surname"\'', $condition->on($user)->toWords());
+        }
+
+        $condition = new BasicCondition('country_id', null);
+
+        $this->assertEquals('Country Id is equal to empty', $condition->on($user)->toWords());
+
+        $condition = new BasicCondition('name', '>', 'Test');
+
+        $this->assertEquals('Name is greater than \'Test\'', $condition->on($user)->toWords());
+
+        $condition = (new BasicCondition('country_id', 2))->negate();
+
+        $this->assertEquals('Country Id is not equal to \'Latvia\'', $condition->on($user)->toWords());
+
+        $condition = new BasicCondition($user->getField('surname'), $user->getField('name'));
+
+        $this->assertEquals('Surname is equal to User Name', $condition->on($user)->toWords());
+
+        $country = clone $this->country;
+
+        $country->addCondition('Users/#');
```

Comment:
need test for `$country->addCondition('#');`

Reply:
`#` is used to set condition on number of referenced records.
What would be the meaning of `$country->addCondition('#');`?

Reply:
topmost table has `#` records

Reply:
`$country->addCondition('Users/#');` - all countries with at least one user.
`$country->addCondition('#');` - countries with at least one country? Doesn't look useful nor spell-able :D

Reply:
> `$country->addCondition('Users/#');` - all countries with at least one user.

I did not looked at the variable name, is 'Users/#' is in reality `country.users`, then

> `$country->addCondition('#');` - countries with at least one country? Doesn't look useful nor spell-able :D

needs to be definitelly supported - and spelling is easy - at least one matching record

Reply:
> at least one matching record

What SQL select it would generate?
`SELECT * FROM country WHERE EXISTS (SELECT 1 FROM country WHERE ...)`
Looks useless.

Reply:
`#` always translate to expr. like `(SELECT COUNT(*) FROM "table before /" WHERE ...)`

so this should answer your question, I updated my answer above - it should either be supported or throw

Reply:
Then throw, because I don't see any use-case for it.

Reply:
Only if it can not be implemented easily, because in trully abstract concept I think it should not require any extra code.